### DISCUSSION TITLE
How determine an assembly is existed in .NET Runtime when building

### DISCUSSION THREAD

**Discussion by Sherlock01** [dotnet/runtime#77769] [Q&A]
We want to block developers from using different .NET Runtime assembly by PackageReference in project. Is there a way to determine an assembly is existed in .NET Runtime.
Can we determine an assembly is .NET Runtime by its PublicKeyToken?

**Comment by Clockwork-Muse** [comment]
> Can we determine an assembly is .NET Runtime by its PublicKeyToken?

Probably, but you'd be making things harder on yourself.

The traditional way to enforce stuff like this is to require your build/dev to use an intercepting package server as the only package source.  These servers usually either allow you to explicitly whitelist packages or package authors or similar (or be completely open).
Note that blocking all third-party packages is probably not reasonable, as opposed to whitelisting a few vetted packages.

**Reply by Sherlock01** [reply]
Thank you for providing the idea.  Our situation is more complex (sorry I didn't make it clear).
We have many repos are using the same package server (feed), and our own packages, parter team packages and 3-part packages are all allowed in the package server.
In single repo, we will not block all projects. We want to block or warn developers when their projects are using .NET Runtime packages (DLLs), and we could allow that if the developer could provide a good reason.
So, detect which projects are using .NET Runtime packages (DLLs) is our best solution.

**Reply by Clockwork-Muse** [reply]
... after that, frankly, your best option is likely code reviews.  If your source control system supports it, require that changes to the `.csproj` requires signoff by whoever can make that sort of determination.  Or who would at least be answerable for signing off; usually the team/tech lead.  Developers should already be limiting the libraries they reference - needing a second approval on top of the package server is usually overkill.  

There exist various dependency audit/reporting tools which could be run during builds, if you're really wanting reporting/traceability (although my google-fu is failing me at the moment).
